arcgis-ios-sdk-samples/Utility Network/Perform valve isolation trace/PerformValveIsolationTraceViewController.swift

```diff
@@ -0,0 +1,271 @@
+        utilityNetwork.features(for: [startingLocationElement]) { [weak self] (features, error) in
+            guard let self = self else { return }
+            if let features = features {
+                if let feature = features.first {
+                    // Get the geometry of the first feature for the starting location as a point.
+                    guard let startingLocationGeometry = feature.geometry as? AGSPoint else {
+                        self.setStatus(message: "Drawing starting location feature failed.")
+                        return
+                    }
+                    // Create a graphic for the starting point and add it to the graphics overlay.
+                    let startingPointSymbol = AGSSimpleMarkerSymbol(style: .cross, color: .green, size: 25)
+                    let startingLocationGraphic = AGSGraphic(geometry: startingLocationGeometry, symbol: startingPointSymbol)
+                    let startingLocationGraphicsOverlay = AGSGraphicsOverlay()
+                    startingLocationGraphicsOverlay.graphics.add(startingLocationGraphic)
+                    self.mapView.graphicsOverlays.add(startingLocationGraphicsOverlay)
+                    self.mapView.setViewpoint(AGSViewpoint(center: startingLocationGeometry, scale: 3000), completion: nil)
+                } else {
+                    self.setStatus(message: "Starting location features not found.")
+                }
+            } else if let error = error {
+                self.setStatus(message: "Loading starting location features failed.")
+                self.presentAlert(error: error)
+            }
+        }
+    }
+    
+    func setStatus(message: String) {
+        statusLabel.text = message
+    }
+    
+    func setUIState() {
+        let utilityNetworkIsReady = utilityNetwork.loadStatus == .loaded
+        categoryButton.isEnabled = utilityNetworkIsReady
+        
+        let canTrace = utilityNetworkIsReady && startingLocationElement != nil
+        traceButton.isEnabled = canTrace
+    }
+    
+    /// Clear all the feature selections from previous trace.
+    func clearLayersSelection() {
+        layers.forEach { $0.clearSelection() }
+    }
+    
+    // MARK: - Actions
+    
+    @IBAction func traceButtonTapped(_ button: UIBarButtonItem) {
+        clearLayersSelection()
+        guard let parameters = makeTraceParameters() else { return }
+        SVProgressHUD.show(withStatus: "Running isolation trace…")
+        utilityNetwork.trace(with: parameters) { [weak self] (traceResults, error) in
+            guard let self = self else { return }
+            if let error = error {
+                self.setStatus(message: "Trace failed.")
+                self.presentAlert(error: error)
+                SVProgressHUD.dismiss()
+                return
+            }
+            guard let elementTraceResult = traceResults?.first as? AGSUtilityElementTraceResult, !elementTraceResult.elements.isEmpty else {
```

Comment:
It seems that this part can be refactored to `if...else` clause, but it will make the code more nested. So I just keep it here as-is.

Reply:
It would cause the code to be indented further, but in this case I think it actually improves readability. The indentation issue can be mitigated by extracting lines `219` through `242` into a separate method:

```swift
func selectFeatures(in elements: [AGSUtilityElement]) { ... }
```